### PR TITLE
compiletest: disable -Aunused for run-pass tests

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Disabled the flag, but that led to quite a bit of fall out -- I think most of it is benign but I've not investigated thoroughly.

r? @petrochenkov

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc https://github.com/rust-lang/rust/issues/43896

**Comment by Centril** [comment]
Please use `#![allow(...)]`s per test instead at least so we can get rid of the noise here.

**Comment by Mark-Simulacrum** [comment]
If we're just going to re-add all the tests to have allow then there's probably no point in doing this, I was just posting this so I didn't forget and for some review on the output. I personally am not sure this is the right approach anymore, as this seems fairly noisy as you said and has a lot of false positives...

**Comment by Centril** [comment]
It's not that many tests really. As for triaging them, I know that at least the warnings in the ATB tests can be safely ignored since I wrote all of those tests.

**Comment by Mark-Simulacrum** [comment]
Not sure I follow you, but I'm not particularly against adding `#![allow(unused)]` to tests where we've triaged them or have prior knowledge of it being fine -- e.g., ATB tests, as you've noted -- I'm just against landing this by re-enabling all run-pass tests as allow(unused) without review.

**Comment by petrochenkov** [comment]
So, my idea was that if a run-pass test has dead code, then it's either
- not a run-pass test, but rather a compile-pass code and should be marked as such, or
- something suspicious is happening and some code intended to run doesn't run.

So, I expected doing triage rather than adding `allow`s or just leaving the `*.stderr` files.
`allow`s can be added with "TRIAGE ME" comments though, so it could be triaged later.

**Comment by Centril** [comment]
@Mark-Simulacrum Oh ofc we should do triage and not stick `allow` there blindly. But I think that should be fairly easy to do for these tests; they are not that many.

**Comment by Mark-Simulacrum** [comment]
I'll add the allows for ATB tests next session I have for impl work; happy to go either route for the rest of the tests.

**Comment by RalfJung** [comment]
> So, my idea was that if a run-pass test has dead code, then it's either
> * not a run-pass test, but rather a compile-pass test and should be marked as such, or
> * something suspicious is happening and some code intended to run doesn't run.

Agreed. In fact, we are specifically [working around](https://github.com/rust-lang/miri/blob/3a5c44c2d83b04651a2e04178087e23e3db6fccc/tests/compiletest.rs#L36) this `-A unued` in Miri because we have had issues where tests had an accidental "unused X" warning that we missed because of this.

**Comment by Mark-Simulacrum** [comment]
Updated. A few tests are moved to check-pass since we don't need to actually run them to verify their results, and otherwise most have allows tagged on.

File: src/test/ui/async-await/issues/issue-54752-async-block.rs [hunk 1]
```diff
@@ -3,4 +3,4 @@
 // edition:2018
 // pp-exact
 
-fn main() { let _a = (async  { }); }
```
Comment (Centril) [inline src/test/ui/async-await/issues/issue-54752-async-block.rs]: Change this back. The parens are what the test is about.

File: src/test/ui/borrowck/borrowck-migrate-to-nll.rs [hunk 2]
```diff
@@ -18,6 +18,8 @@
 //[zflag] run-pass
 //[edition] run-pass
```
Comment (Centril) [inline src/test/ui/borrowck/borrowck-migrate-to-nll.rs]: Can be lowered to build-pass at minimum.

File: src/test/ui/consts/const-labeled-break.rs [hunk 3]
```diff
@@ -5,6 +5,8 @@
 //
 // See https://github.com/rust-lang/rust/issues/51350 for more information.
 
+#![allow(dead_code, unreachable_code)]
+
 const CRASH: () = 'a: while break 'a {};
```
Comment (Centril) [inline src/test/ui/consts/const-labeled-break.rs]: Can be build-pass at minimum.

**Comment by Mark-Simulacrum** [comment]
Fixed review comments.

File: src/test/ui/async-await/multiple-lifetimes/hrtb.rs [hunk 4]
```diff
@@ -1,10 +1,11 @@
 // edition:2018
-// run-pass
+// check-pass
 
 // Test that we can use async fns with multiple arbitrary lifetimes.
 
 use std::ops::Add;
 
+#[allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/async-await/multiple-lifetimes/hrtb.rs]: Unnecessary in `check-pass`.

File: src/test/ui/borrowck/borrowck-migrate-to-nll.rs [hunk 5]
```diff
@@ -15,8 +15,10 @@
 // revisions: zflag edition
 //[zflag]compile-flags: -Z borrowck=migrate
 //[edition]edition:2018
-//[zflag] run-pass
-//[edition] run-pass
+//[zflag] check-pass
+//[edition] check-pass
+
+#![allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/borrowck/borrowck-migrate-to-nll.rs]: Unnecessary in `check-pass`.

File: src/test/ui/const-generics/apit-with-const-param.rs [hunk 6]
```diff
@@ -3,6 +3,8 @@
 #![feature(const_generics)]
 //~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
 
+#![allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/const-generics/apit-with-const-param.rs]: No a `run-pass` test.

File: src/test/ui/const-generics/issue-61422.rs [hunk 7]
```diff
@@ -3,12 +3,14 @@
 #![feature(const_generics)]
 //~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
 
+#![allow(unused_variables, dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/const-generics/issue-61422.rs]: No a `run-pass` test.

File: src/test/ui/const-generics/unused-const-param.rs [hunk 8]
```diff
@@ -3,6 +3,8 @@
 #![feature(const_generics)]
 //~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
 
+#![allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/const-generics/unused-const-param.rs]: No a `run-pass` test.

File: src/test/ui/deprecation/deprecation-in-future.rs [hunk 9]
```diff
@@ -1,10 +1,11 @@
-// run-pass
+// check-pass
 
 #![deny(deprecated_in_future)]
 
 #[deprecated(since = "99.99.99", note = "text")]
 pub fn deprecated_future() {}
 
+#[allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/deprecation/deprecation-in-future.rs]: Unnecessary in `check-pass`.

File: src/test/ui/hrtb/issue-57639.rs [hunk 10]
```diff
@@ -20,6 +20,7 @@ trait Bar {
     type Item: for<'a> Foo<'a>;
 }
 
+#[allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/hrtb/issue-57639.rs]: Unnecessary in `check-pass`.

File: src/test/ui/impl-trait/closure-calling-parent-fn.rs [hunk 11]
```diff
@@ -5,7 +5,9 @@
 // `foo` and hence is treated opaquely within the closure body.  This
 // resulted in a failed subtype relationship.
 //
-// run-pass
+// check-pass
+
+#![allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/impl-trait/closure-calling-parent-fn.rs]: Unnecessary in `check-pass`.

File: src/test/ui/issues/issue-61711-once-caused-rustc-inf-loop.rs [hunk 12]
```diff
@@ -7,5 +7,6 @@
 
 // run-pass
 
+#[allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/issues/issue-61711-once-caused-rustc-inf-loop.rs]: Not a `run-pass` test.

**Comment by Mark-Simulacrum** [comment]
Force-pushed fixes.

**Comment by Mark-Simulacrum** [comment]
@bors r-

Rebase error, misunderstood test results locally.

**Pull request merged**